### PR TITLE
CommonJS support

### PR DESCRIPTION
**Pull request by bartlomieju**
Adds "CommonJS" support section to "Node compatibility" page.

### CODE REVIEW

File: runtime/fundamentals/node.md [hunk 1]
```diff
@@ -56,6 +56,175 @@ error: Relative import path "os" not prefixed with / or ./ or ../
 Same hints and additional quick-fixes are provided by the Deno LSP in your
 editor.
 
+## CommonJS support
+
+CommonJS is a module system that predetes
```
Comment (thisisjofrank) [inline runtime/fundamentals/node.md:61]: predates? Or preceded?

File: runtime/fundamentals/node.md [hunk 2]
```diff
@@ -56,6 +56,175 @@ error: Relative import path "os" not prefixed with / or ./ or ../
...
+[ES modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules).
+While we firmly believe that ES modules are the future of JavaScript, there are
+millions of npm libraries that are written in CommonJS and Deno offers full
+support for them. Deno will automatically determine if a package is using
+CommonJS and makes it work seemlessly when imported:
```
Comment (thisisjofrank) [inline runtime/fundamentals/node.md:66]:
```suggestion
CommonJS and make it work seamlessly when imported:
```

File: runtime/fundamentals/node.md [hunk 3]
```diff
@@ -56,6 +56,175 @@ error: Relative import path "os" not prefixed with / or ./ or ../
...
+import react from "npm:react";
+console.log(react);
+```
+
+```shell
+$ deno run -E main.js
+18.3.1
+```
+
+_`npm:react` is a CommonJS package. Deno provides a seemless way to import it._
```
Comment (thisisjofrank) [inline runtime/fundamentals/node.md:78]:
```suggestion
_`npm:react` is a CommonJS package. Deno provides a seamless way to import it._
```

Comment (thisisjofrank) [reply runtime/fundamentals/node.md:78]: Duplication of the word "seamless" from above reads a little oddly.

File: runtime/fundamentals/node.md [hunk 4]
```diff
@@ -56,6 +56,175 @@ error: Relative import path "os" not prefixed with / or ./ or ../
...
+
+That said, Deno strongly encourages use of ES modules, and offers restriced
+support for CommonJS in user code. There are two ways to write and consume
```
Comment (thisisjofrank) [inline runtime/fundamentals/node.md:81]:
```suggestion
support for CommonJS in user code.
 
There are two ways to write and consume
```

File: runtime/fundamentals/node.md [hunk 5]
```diff
@@ -56,6 +56,175 @@ error: Relative import path "os" not prefixed with / or ./ or ../
...
+
+### Use `.cjs` extension
+
+If the file extension is `.cjs` Deno will treat this module as CommonJS.
+
+```js title="main.cjs"
+const express = require("express");
+```
+
+Deno does not look for `package.json` files and `type` option to determine if
```
Comment (thisisjofrank) [inline runtime/fundamentals/node.md:92]:
```suggestion
Deno does not look for `package.json` files or `type` options to determine if
```

File: runtime/fundamentals/node.md [hunk 6]
```diff
@@ -56,6 +56,175 @@ error: Relative import path "os" not prefixed with / or ./ or ../
...
+the file is CommonJS or ESM.
+
+When using CommonJS, Deno expects that dependencies will be installed manually
+and `node_modules` directory will be present. It's best to set
```
Comment (thisisjofrank) [inline runtime/fundamentals/node.md:96]:
```suggestion
and a `node_modules` directory will be present. It's best to set
```

File: runtime/fundamentals/node.md (hunk 5 above)
Comment (bartlomieju) [reply runtime/fundamentals/node.md:92]: I believe this is actually correct - `type` is only specified in `package.json`. So you can't have `type` without `package.json` hence the "and"

File: runtime/fundamentals/node.md [hunk 7]
```diff
@@ -56,6 +56,175 @@ error: Relative import path "os" not prefixed with / or ./ or ../
...
+[Function: createApplication] {
+  application: {
+    init: [Function: init],
+    defaultConfiguration: [Function: defaultConfiguration],
+    ...
+  }
+}
+```
+
+Notice that `-R` and `-E` flags were used to give access to read and env
```
Comment (thisisjofrank) [inline runtime/fundamentals/node.md:118]:
```suggestion
`-R` and `-E` flags are used to allow permissions to read files and environment variables.
```

File: runtime/fundamentals/node.md [hunk 8]
```diff
@@ -56,6 +56,175 @@ error: Relative import path "os" not prefixed with / or ./ or ../
...
+permissions.
+
+**Deno permission system is still in effect when using CommonJS modules.** It is
```
Comment (thisisjofrank) [inline runtime/fundamentals/node.md:121]:
```suggestion
**Deno's permission system is still in effect when using CommonJS modules.** It is
```

File: runtime/fundamentals/node.md [hunk 9]
```diff
@@ -56,6 +56,175 @@ error: Relative import path "os" not prefixed with / or ./ or ../
...
+necessary to provide at least `--allow-read` permission as Deno will probe the
+file system for `package.json` files and `node_modules` directory to properly
+resolve CommonJS modules.
+
+### Create `require()` manually
+
+Alternative option is to create an instance of `require()` function manually,
```
Comment (thisisjofrank) [inline runtime/fundamentals/node.md:128]:
```suggestion
An alternative option is to create an instance of the `require()` function manually,
```

File: runtime/fundamentals/node.md [hunk 10]
```diff
@@ -56,6 +56,175 @@ error: Relative import path "os" not prefixed with / or ./ or ../
...
+this can be done as such:
```
Comment (thisisjofrank) [inline runtime/fundamentals/node.md:129]:
```suggestion
```

**Review by marvinhagemeister — APPROVED** [review]
LGTM

**Pull request merged**